File: 2020/09/0906/세그먼트트리.py
import sys
from math import log, ceil


def init(tree, board, node, start, end):
    if start == end:
        tree[node] = board[start]
    else:
        init(tree, board, node*2, start, (start+end)//2)
        init(tree, board, node*2+1, (start+end)//2+1, end)
        tree[node] = min(tree[node * 2], tree[node * 2 + 1])


def query(tree, node, start, end, i, j):
    if i > end or j < start:
        return -1
    if i <= start and end <= j:
        return tree[node]
    m1 = query(tree, 2*node, start, (start+end)//2, i, j)
    m2 = query(tree, 2*node+1, (start+end)//2+1, end, i, j)
    if m1 == -1:
        return m2
    elif m2 == -1:
        return m1
    else:
        return min(m1, m2)


sys.stdin = open("10868.txt", "r")
N, M = map(int, input().split())
H = ceil(log(N, 2))
size = (1 << (H+1))
board = [int(input()) for _ in range(N)]
tree = [0] * size

init(tree, board, 1, 0, N-1)
for _ in range(M):
    start, end = map(int, input().split())
    print(query(tree, 1, 0, N-1, start-1, end-1))

File: 자스베이스/구간트리/세그먼트트리/세그먼트트리.js
// const input = require('fs').readFileSync('/dev/stdin').toString().split("\n");
// const input = require('fs').readFileSync('dev/stdin').toString().split('\n').map(x=>x.split(' ').map(x=>x.trim()));

const input = require("fs").readFileSync("./10868.txt").toString().split("\n");
const [N, M] = input.shift().split(" ").map(Number);
const board = input.splice(0, N).map((x) => parseInt(x));
const log2 = 1 << (Math.ceil(Math.log2(N)) + 1);
const tree = Array(log2).fill(0);
const INF = Number.MAX_SAFE_INTEGER;

const init = (x, s, e) => {
  if (s === e) {
    tree[x] = board[e];
    return tree[x];
  }
  mid = parseInt((s + e) / 2);
  tree[x] = Math.min(init(2 * x, s, mid), init(2 * x + 1, mid + 1, e));
  return tree[x];
};

init(1, 0, N - 1);
const query = (x, s, e, S, E) => {
  if (S > e || s > E) return INF;
  if (s >= S && e <= E) return tree[x];
  mid = parseInt((s + e) / 2);
  return Math.min(
    query(2 * x, s, mid, S, E),
    query(2 * x + 1, mid + 1, e, S, E)
  );
};

for (let temp of input) {
  const [start, end] = temp.split(" ").map((x) => parseInt(x));
  console.log(query(1, 0, N - 1, start - 1, end - 1));
}
